### ISSUE TITLE
zsh completion is mistaken

### ISSUE THREAD

**Issue by WhyNotHugo** [pimutils/khal#1062]
```
$ khal list --<tab>
--days              -- specify how many days to include
--events            -- specify how many events to include
--exclude-calendar  -- don't use this calendar
--help              -- show help information
--include-calendar  -- specify calendar to use
$ khal list --days 20
Usage: khal list [OPTIONS] [DATETIME [DATETIME | RANGE]]
Try 'khal list --help' for help.

Error: No such option: --days

**Comment by sdx23** [comment]
I began taking on this by manually editing `misc/__khal`, realized it's a quite boring task, considered using https://github.com/RobSis/zsh-completion-generator and finally thought one should fix this "upstream" i.e. in click.

Turns out click can automatically generate completions: https://github.com/pallets/click/tree/main/examples/completion
The result somewhat works. It can complete subcommands and options, but not both; neither does it do parameter for subcommands. Possibly this may be improved somewhat by adding decorators and/or extending existing ones.

However, I'm not sure whether that's the way to go or we should rather write an up-to-date __khal. The optimal solution imho would be if click could generate a _khal definition that doesn't rely on repeated calls to khal.

**Comment by WhyNotHugo** [comment]
I'd prefer to do as much as possible using what upstream has to offer, and eventually submit improvements there.

Having custom made completion adds a bit of maintenance burden. Manually made completion needs to be updated every time something changes, whereas something generated by click should update itself.

**Comment by sdx23** [comment]
I agree. However, I can estimate neither how difficult the required changes upstream to click would be nor whether they're possible at all.

To clarify what I meant by "that doesn't rely on repeated calls to khal", the attachement "khal1.txt" is what the first invocation `_KHAL_COMPLETE=zsh_source khal` outputs. The contained `env COMP_WORDS="${words[*]}" COMP_CWORD=$((CURRENT-1)) _KHAL_COMPLETE=zsh_complete khal` is another call to khal which outputs "khal2.txt". Due to the structuring of khal in commands and subcommands with their very own options this may become rather suboptimal from an efficiency point of view.

[khal2.txt](https://github.com/pimutils/khal/files/6786792/khal2.txt)
[khal1.txt](https://github.com/pimutils/khal/files/6786794/khal1.txt)

**Comment by WhyNotHugo** [comment]
I understand your point, but don't have enough understanding of autocompletion mechanics to comment further.

Can we get away with very few calls to `khal` using click's completion framework?

**Comment by sdx23** [comment]
Having fewer calls (e.g. by inserting a fixed commands list into the completion script) would result in a higher maintenance burden. You cannot have both.

To fully switch to clicks completion, we'd just need to ship the content of "khal1.txt" in https://github.com/pimutils/khal/blob/master/misc/__khal
However, imho completion is not fixed by that and I therefore didn't propose this as a PR.

List of stuff to fix (incomplete):
- [ ] -v has fixed list of possible values "CRITICAL, ERROR, WARNING, INFO or DEBUG"
- [ ] -a should propose calendars from config
- [ ] -d dito
- [ ] list command: propose [DATETIME [DATETIME | RANGE]]
- [ ] at command: propose [[START DATE] TIME | now]
- [ ] calendar command: propose [DATERANGE]
- [ ] edit command: propose [SEARCH_STRING]
- [ ] new command: propose  [START [END | DELTA] [TIMEZONE] [SUMMARY] [::DESCRIPTION]]
- [ ] new command: options
- [ ] printics command: complete ics files only
- [ ] formatting options: propose allowed formats

**Comment by geier** [comment]
closed with #1141
